src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -1879,21 +2102,54 @@ case class IndexArray(posn: Position, f: AST, idx: AST) extends AST(posn, Array(
 
   def eval(ec: EvalContext): () => Any = ((f.`type`, idx.`type`): @unchecked) match {
     case (t: TArray, TInt) =>
+      val localT = t
+      val localPos = posn
       AST.evalCompose[IndexedSeq[_], Int](ec, f, idx)((a, i) =>
         try {
-        a(i)
+          a(i)
         } catch {
           case e: java.lang.IndexOutOfBoundsException =>
-            fatal(s"Tried to access index [$i] on array ${compact(t.makeJSON(a))} of length ${a.length}" +
+            ParserUtils.error(localPos,
+              s"Tried to access index [$i] on array ${compact(localT.makeJSON(a))} of length ${a.length}" +
               s"\n  Hint: All arrays in Hail are zero-indexed (`array[0]' is the first element)" +
               s"\n  Hint: For accessing `A'-numbered info fields in split variants, `va.info.field[va.aIndex]' is correct")
           case e: Throwable => throw e
         })
 
+    case (TDict(_), TString) =>
+      AST.evalCompose[Map[_, _], String](ec, f, idx)((d, k) =>
+        d.asInstanceOf[Map[String, _]]
+          .get(k)
+          .orNull
+      )
+
     case (TString, TInt) =>
       AST.evalCompose[String, Int](ec, f, idx)((s, i) => s(i).toString)
   }
+}
+
+case class SliceArray(posn: Position, f: AST, idx1: Option[AST], idx2: Option[AST]) extends AST(posn, Array(Some(f), idx1, idx2).flatten) {
+  override def typecheckThis(): BaseType = f.`type` match {
+    case (t: TArray) =>
+      if (idx1.exists(_.`type` != TInt) || idx2.exists(_.`type` != TInt))
+        parseError(s"invalid slice expression.  " +
+          s"Expect (array[start:end] || array[:end] || array[start:]) where start and end are integers, " +
+          s"but got [${idx1.map(_.`type`).getOrElse("")}:${idx2.map(_.`type`).getOrElse("")}]")
+      else
+        t
+    case _ => parseError(s"invalid slice expression.  Only arrays can be sliced, tried to slice type `${f.`type`}'")
+  }
 
+  def eval(ec: EvalContext): () => Any =
+    (idx1, idx2) match {
+      case (Some(i1), Some(i2)) =>
+        AST.evalCompose[IndexedSeq[_], Int, Int](ec, f, i1, i2)((a, ind1, ind2) => a.slice(ind1, ind2))
+      case (Some(i1), None) =>
+        AST.evalCompose[IndexedSeq[_], Int](ec, f, i1)((a, ind1) => a.slice(ind1, a.length))
+      case (None, Some(i2)) =>
+        AST.evalCompose[IndexedSeq[_], Int](ec, f, i2)((a, ind2) => a.slice(0, ind2))
+      case (None, None) => f.eval(ec)
```

Comment:
Yes, you're right.  There are a few ways around this but I don't think they are worth it.